### PR TITLE
Added initial draft of NEAT algorithm post

### PR DESCRIPTION
**Pull request by thinkmorestupidless**

### CODE REVIEW

**Review by PeuTit — APPROVED** [review]
Great Read on a very fascinating subject! LGTM expect one small typo.

**Review by PeuTit — APPROVED** [review]

**Review by gianick — APPROVED** [review]

File: media/2024-02-29-the-neat-algorithm-evolving-neural-network-topologies/picbreeder-results.png
[the remark this answers is not in the document]
Comment (thinkmorestupidless) [reply media/2024-02-29-the-neat-algorithm-evolving-neural-network-topologies/picbreeder-results.png:1]: I don't see any border, let's see what it looks like on acceptance

File: posts/2024-02-29-the-neat-algorithm-evolving-neural-network-topologies.adoc [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+
+I don't propose we should scour the literature for the next paradigm shift - the next link:https://en.wikipedia.org/wiki/Dynabook[DynaBook] or link:https://en.wikipedia.org/wiki/Project_Xanadu[Xanadu]. I wouldn't ignore it if I were to stumble across it, of course, but there is a huge amount of utility in looking for the clever techniques which have solved difficult problems on a less grand scale. You never know what you'll find, and you never know what will prove useful.
+
+Doesn't that sound more fun than spending hours crafting just the right prompt?
+
+== Neuro-Evolution of Augmenting Topologies (NEAT)
+
+In this post I'd like to share something that's perhaps not exactly a forgotten relic, but certainly something that's not very well known - the NEAT algorithm.
```
[the remark this answers is not in the document]
Comment (thinkmorestupidless) [reply posts/2024-02-29-the-neat-algorithm-evolving-neural-network-topologies.adoc:26]: Well i'm describing NEAT in the blog post so i'm going to leave that one

File: posts/2024-02-29-the-neat-algorithm-evolving-neural-network-topologies.adoc [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+
+This is what NEAT does, it is an algorithm which creates a neural network that solves whatever problem you define.
+
+To be more precise, NEAT is a link:https://en.wikipedia.org/wiki/Genetic_algorithm[Genetic Algorithm] which evolves the least complex network topology capable of approximating a target function. The neural networks NEAT outputs are capable of displaying extremely complex behaviours.
+
+Rather than starting with a fixed topology created by a human, NEAT starts with a simple neural network topology (for instance, each input node connected to a single output node) and a fitness function and iteratively adds structure to the topology (i.e. adding a new node or a new connection between existing nodes, perturbing the weights of nodes, changing the activation function of a node, etc., etc.) until the output of the network satisfies the error threshold of the fitness function.
+
+And if you don't believe me about the Mario thing, here's a video.
+
+video::qv6UVOQ0F44[youtube,width=960,height=540]
```
[the remark this answers is not in the document]
Comment (thinkmorestupidless) [reply posts/2024-02-29-the-neat-algorithm-evolving-neural-network-topologies.adoc:47]: It's introduced in the line above so i'll leave it out of this one

File: posts/2024-02-29-the-neat-algorithm-evolving-neural-network-topologies.adoc [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+
+===== Compatibility
+
+Categorizing organisms into groups which share similar topologies is another problem for which innovation numbers provide an elegant solution.
+
+When we compare the genomes of two organisms the more disjoint and excess genes there are the less evolutionary history they share, meaning they're less compatible. In practice, it's not just disjoint and excess genes which are counted but also a measure of the average weight differences between matching genes.
+
+The NEAT paper encodes the formula for compatibility in this magical spell:
+
+image:compatibility-function.png[]
```
[the remark this answers is not in the document]
Comment (thinkmorestupidless) [reply posts/2024-02-29-the-neat-algorithm-evolving-neural-network-topologies.adoc:187]: Yeah, it's generated by latex and not great, let's see what it looks like on acceptance and if it's still not good i'll draw it up in Photoshop or something.

File: posts/2024-02-29-the-neat-algorithm-evolving-neural-network-topologies.adoc [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+
+== Why Use NEAT?
+
+We've covered how NEAT works, and it's cool and everything, but why would you use it rather than just copy/pasting some PyTorch code from ChatGPT? What's the trade-off for the extra complexity involved? Is it better than anything else for solving any particular class of problem?
+
+.This is the conclusion from the original NEAT paper:
+[quote,Efficient Reinforcement Learning through Evolving Neural Network Topologies,Stanley & Miikkulainen 2002]
+...evolving structure and connection weights in the style of NEAT leads to significant performance gains in reinforcement learning. NEAT exploits
+properties of both structure and history that have not been
+utilized before. Historical markings, protection of innovation through speciation, and incremental growth from minimal structure result in a system that is capable of evolving solutions of minimal complexity. NEAT is a unique
```
[the remark this answers is not in the document]
Comment (thinkmorestupidless) [reply posts/2024-02-29-the-neat-algorithm-evolving-neural-network-topologies.adoc:203]: This was a copy/paste from the paper, and as it's in a block quote i'm going to leave it.

File: posts/2024-02-29-the-neat-algorithm-evolving-neural-network-topologies.adoc [hunk 5]
```diff
@@ -0,0 +1,401 @@
+= The NEAT Algorithm: Evolving Neural Networks
+thinkmorestupidless
+v1.0, 2024-02-29
+:title: The NEAT Algorithm: Evolving Neural Networks
+:imagesdir: ../media/2024-02-29-the-neat-algorithm-evolving-neural-network-topologies
+:lang: en
+:tags: [artifical intelligence, machine learning, algorithm, machine learning]
```
Comment (gianick) [inline posts/2024-02-29-the-neat-algorithm-evolving-neural-network-topologies.adoc:7]:
```suggestion
:tags: [artifical intelligence, machine learning, algorithm]
```

**Review by gianick — APPROVED** [review]

**Pull request merged**